sw/device/silicon_creator/lib/drivers/uart.c

```diff
@@ -9,101 +9,90 @@
   }
 
-  // Calculation formula: NCO = 16 * 2^nco_width * baud / fclk.
-  // NCO creates 16x of baudrate. So, in addition to the nco_width,
-  // 2^4 should be multiplied.
-  uint64_t nco =
-      ((uint64_t)uart->baudrate << (NCO_WIDTH + 4)) / uart->clk_freq_hz;
-  uint32_t nco_masked = nco & UART_CTRL_NCO_MASK;
-
-  // Requested baudrate is too high for the given clock frequency.
-  if (nco != nco_masked) {
-    return kErrorUartBadBaudRate;
-  }
-
   // Must be called before the first write to any of the UART registers.
-  uart_reset(uart);
+  uart_reset();
 
   // Set baudrate, TX, no parity bits.
   uint32_t reg = 0;
-  reg = bitfield_field32_write(reg, UART_CTRL_NCO_FIELD, nco_masked);
+  reg = bitfield_field32_write(reg, UART_CTRL_NCO_FIELD, precalculated_nco);
   reg = bitfield_bit32_write(reg, UART_CTRL_TX_BIT, true);
   reg = bitfield_bit32_write(reg, UART_CTRL_PARITY_EN_BIT, false);
-  mmio_region_write32(uart->base_addr, UART_CTRL_REG_OFFSET, reg);
+  abs_mmio_write32(TOP_EARLGREY_UART0_BASE_ADDR + UART_CTRL_REG_OFFSET, reg);
 
   // Disable interrupts.
-  mmio_region_write32(uart->base_addr, UART_INTR_ENABLE_REG_OFFSET, 0u);
+  abs_mmio_write32(TOP_EARLGREY_UART0_BASE_ADDR + UART_INTR_ENABLE_REG_OFFSET,
+                   0u);
   return kErrorOk;
 }
 
-static bool uart_tx_full(const uart_t *uart) {
-  uint32_t reg = mmio_region_read32(uart->base_addr, UART_STATUS_REG_OFFSET);
+static bool uart_tx_full(void) {
+  uint32_t reg =
+      abs_mmio_read32(TOP_EARLGREY_UART0_BASE_ADDR + UART_STATUS_REG_OFFSET);
   return bitfield_bit32_read(reg, UART_STATUS_TXFULL_BIT);
 }
 
-static bool uart_tx_idle(const uart_t *uart) {
-  uint32_t reg = mmio_region_read32(uart->base_addr, UART_STATUS_REG_OFFSET);
+static bool uart_tx_idle(void) {
+  uint32_t reg =
+      abs_mmio_read32(TOP_EARLGREY_UART0_BASE_ADDR + UART_STATUS_REG_OFFSET);
   return bitfield_bit32_read(reg, UART_STATUS_TXIDLE_BIT);
 }
 
-void uart_putchar(const uart_t *uart, uint8_t byte) {
+void uart_putchar(uint8_t byte) {
   // If the transmit FIFO is full, wait.
-  while (uart_tx_full(uart)) {
+  while (uart_tx_full()) {
   }
   uint32_t reg = bitfield_field32_write(0, UART_WDATA_WDATA_FIELD, byte);
-  mmio_region_write32(uart->base_addr, UART_WDATA_REG_OFFSET, reg);
+  abs_mmio_write32(TOP_EARLGREY_UART0_BASE_ADDR + UART_WDATA_REG_OFFSET, reg);
 
   // If the transmitter is active, wait.
-  while (!uart_tx_idle(uart)) {
+  while (!uart_tx_idle()) {
   }
 }
 
 /**
  * Write `len` bytes to the UART TX FIFO.
  */
-size_t uart_write(const uart_t *uart, const uint8_t *data, size_t len) {
+size_t uart_write(const uint8_t *data, size_t len) {
   size_t total = len;
   while (len) {
-    uart_putchar(uart, *data);
+    uart_putchar(*data);
     data++;
     len--;
   }
   return total;
 }
 
 size_t uart_sink(void *uart, const char *data, size_t len) {
-  return uart_write((const uart_t *)uart, (const uint8_t *)data, len);
+  (void)uart;
+  return uart_write((const uint8_t *)data, len);
```

Comment:
Why not static_cast? Are old-style casts allowed by the style guide?

Reply:
C vs C++.  This is C code, so there is no static_cast.

Reply:
Whoa... I was not aware we still had C around :) Got it.